### PR TITLE
python3 compatibility changes

### PR DESCRIPTION
**Pull request by tasleson**
This is a pretty invasive change and it's ~~not~~ mostly complete.  These changes should allow the code to run on both python2 & python3.  List of items that need to be resolved (help appreciated):

1. ~~`check_const.pl` needs to be updated to support the py3 change (commented out at the moment)~~
2. ~~`argparse` is a mess, see commit 75f82010fc426bce5f76223f7522b8343ccccfe0 for details~~ Removed with rebased code and things are working
3. `simarray.py` has a bug with `sim_pool['free_space']` (we have a workaround in place, need to get to root cause).
4. ~~More work on smis plugin in error paths to ensure error paths are same from lmiwbem~~

Missing packages for Travis CI:
1. ~~python3-lmiwbem~~ Mitigated by supporting pywbem too
2. ~~python-future~~ This requirement has been removed

~~In addition we need to figure out how to handle `#!/bin/env python2` vs. using  `#!/bin/env python3` in those python files that have it.   Originally we used  `#!/bin/env python` and that raised an issue when packaging on rpm based distros as they want the files to be specific to a python version.  However, in this case if the code can and does work for both I don't see why we couldn't change back to being less specific.  That would make packaging a lot easier, otherwise we need to edit files during packaging.~~

~~From a pure packaging standpoint we are going to need to prevent people from installing a python2 and python3 version of the same package if the package supplies a file in common, eg. lsmcli or one of the plugins.~~ We are following packaging rules for the platform.

The following plugins are tested and passing with py2 & py3
* simc/sim (automated)
* nstor (automated)
* targetd (automated)
* ontap (automated, added 10/18)
* smispy (automated, Note: lmi-wbem not working with SSL)
* hpsa (automated, **not** always available at this point in time)

Need to be tested
* megaraid

### CODE REVIEW

**Comment by tasleson** [comment]
Yep, I should have remembered to update the lab to add dependencies and update the travis file too.

**Comment by tasleson** [comment]
Re-based the branch, no testing done and **a lot** of merge conflicts were needed.

**Comment by tasleson** [comment]
@cathay4t I'm passing `make check` when running with python2 and python3

**Comment by cathay4t** [comment]
Some thoughts:
 * Currently python plugins are forced to use python2, how about we introduce `configure.ac` options
    like  `./configure --plugin-python-version 3|2|default` which generates:
   * `#!/usr/bin/env python2`
   * `#!/usr/bin/env python3`
  * `#!/usr/bin/env python`

 * No need to force the use of `lmiwbem` as `pywbem >= 0.8` supports python3 also.
   Tested on pywbem 0.8.4 with smispy plugin in python2 and python3 mode.
   I will try to upgrade  pywbem Fedora.
 * Please remove the use of `python-future`:
   * ONTAP plugin does not require `from future import standard_library`. Tested.
   * The `six.with_metaclass` could replace `from future.utils import with_metaclass`. Tested.
   * The `python-future` is EPEL7 package while libstoragemgmt is RHEL7 package.

**Comment by tasleson** [comment]
Thanks for the comments @cathay4t .  I was not planning on making python3 support work for EL7/RHEL7, that will just stay with py2 support.

**Comment by cathay4t** [comment]
@tasleson Could you include this [patch][1] to support both pywbem and lmiwbem?

Still working on `./configure --enable-py3-plugin` and `./configure --enable-py3-cli`.

[1]: https://github.com/cathay4t/libstoragemgmt/commit/072f9e6cc5bed109ed248fa733ae98819cd2b5b2.patch

**Comment by tasleson** [comment]
@cathay4t Patch added as requested

**Comment by tasleson** [comment]
@cathay4t Rebased, I reworked the patch set and made it simpler too (I hope :-)

**Comment by tasleson** [comment]
`make check` for py2 and py3 work.  Nstor & targed only tested with py2.  We need to devise a way to run the source code against py2 & py3 and across all plugins.

**Comment by cathay4t** [comment]
@tasleson I add some patches for this in PR #193 
Please try to pull them  and merge into this PR.

Thank you.

**Comment by tasleson** [comment]
@cathay4t Changes from PR #193 pulled in, thanks!

**Comment by tasleson** [comment]
@cathay4t Please modify `check_const.pl` to work with py3 changes.

**Comment by cathay4t** [comment]
@tasleson The 'check_const.pl' should works if you include 'Python: Use absolute import as PEP8 suggested.'

**Comment by tasleson** [comment]
@cathay4t OK, I've re-integrated your patches where I believe it's appropriate.  Let me know if this is closer to what you were thinking.  The only actual content difference is:

```diff
diff --git a/test/runtests.sh.in b/test/runtests.sh.in
index c60b98a..a5815e4 100644
--- a/test/runtests.sh.in
+++ b/test/runtests.sh.in
@@ -38,7 +38,7 @@ with_mem_leak_test="@WITH_MEM_LEAK_TEST@"
 source "${src_dir}/test/test_include.sh"
 
 # Constant check: check whether python constants matched with C constants.
-# perl ${src_dir}/tools/utility/check_const.pl || exit 1
+perl ${src_dir}/tools/utility/check_const.pl || exit 1
 
 # First round: test against python plugin sim://
 lsm_test_base_install \

```

**Comment by tasleson** [comment]
OK, so I've been playing around with trying to test this some more.  I'm open for suggestions on how we can easily:

1. Test py2 client with py2 plugins
2. Test py2 client with py3 plugins
3. Test py3 client with py3 plugins

With the current changes in this branch we can test 1 & 3, but not 2.  I'm open for suggestions on how we can achieve this test matrix above as easily and reliably as possible and make it work within the CI infrastructure against all the available hardware.

Previously I had the shared library named differently for py3 which allowed both shared libraries to exist in the same directory and I just used: 

```perl
perl -pi -e "s/env python2/env python3/"
```
with `find` to switch from one to the other.  Latest changes have the shared lib named the same which works fine for deployment as they are in different directories.

Tripling our test matrix and all the associated hassle of supporting py2 & py3 and the combination of each has made me rethink the value of utilizing python as the `"cost"` of using python continues to add up.

**Comment by joehandzik** [comment]
@tasleson Well, I can tell you that python support is a big value-add for us (two different solutions that I have in-flight rely on the python API, Ceph will be the only C API user). There's another potential python user in play too, so I think the python is an absolute must-have for us at HPE going forward...not that it isn't a pain, just that it's definitely worth it for my partner use cases.

As for the testing matrix...personally, I wonder...do we actually need to test all three? Or can we reasonably say that the client and plugin python version need to match? At a certain point, the test matrix is too much of a pain to be sustainable. I think testing matched client/plugin versions would be a reasonable best effort.

**Comment by tasleson** [comment]
@joehandzik Don't worry we will continue to support python, that's not a question especially on the client side.  This was just more of me venting and thinking that in the future I will not select python unless it's a requirement for something.  Personally I don't think enough gains were made with the py3 language change to justify the incredible amount of effort it takes to support both.  They should have just incrementally improved py2 IMHO.  I'm seriously looking at go, rust and others for new future development instead.  We've also looked at supporting dbus and just let clients write their own dbus client side code too, in whatever language they write in.

As for the testing matrix I think maybe we can skip 2 above on real hardware as long as we test that combination with the simulator(s).  We need to make sure that a py2 client can talk to py3 backend to ensure the IPC between the two of them works as expected.

However, running the tests 1 & 3 on real hardware is ugly enough, but I don't see a way around it unless we re-write the plugins in a different language.  This is something that is technically possible, but not something I really want to do.

**Comment by cathay4t** [comment]
In current code base, make check with `with-python3` configure option will test python3 client with python3 plugin. By default, make check check python2 client with python2 plugin.
I will try to find a way to test python2 client with python3 plugin.

**Comment by tasleson** [comment]
@cathay4t Make check only works for simulated plugins.  We still need to run all the other plugins with py2 & py3 to ensure correctness.

**Comment by joehandzik** [comment]
@tasleson Makes sense, seeing the effort Red Hat as a company is going to in order to actually support python3 consistently has definitely called into question the value of it. If a similar jump between python versions is ever required again, I figure folks really will just migrate to one of the other languages you mentioned. As a side note, I've wondered if it would be worthwhile to rearchitect the core of libstoragemgmt as Go. Maybe something worth investigating.

I think your proposal for testing option 2 would be adequate. 

And agreed that we shouldn't rewrite the plugins in a new language, at least not right now.

**Comment by tasleson** [comment]
Nstor plugin reworked as requests library was not working with v4 nstor.

**Comment by tasleson** [comment]
@cathay4t The lmiwbem ssl support for python3 is returning: 
`lmiwbem_core.ConnectionError: (45, 'Cannot connect to 10.1.134.227:5989. Connection failed.')`  Pywbem for py2 works fine to same box.

I thought I tested this early on py3 testing, but I can't recall.  You were going to look at pywbem for py3 and possibly package that for fedora, any success on that?

**Comment by tasleson** [comment]
@cathay4t Ugh, looks like pywbem changed their API, WRT to error handling class names.  Looks like `pywbem.cim_http.Error` may have been changed to `pywbem.cim_http.HTTPError`.  If this is indeed true I'm going to have yet another library that I'm disappointed in.

**Comment by cathay4t** [comment]
@tasleson I only got pywbem py3 version works on ArchLinux. Will look into the issues above tomorrow and create bug(with rpm spec patch) for upgrae pywbem in Fedora.

**Comment by tasleson** [comment]
@cathay4t SMI-S is working with one exception, SSL on py3 using lmiwbem.  The newer versions of pywbem which support py3 have changes that will need to be accounted for if we end up going that route.  I believe lmiwbem is in need of a new maintainer and I'm not sure who is going to pick that up, so that may push us more towards using newer versions of pybem and dealing with all the issues that will create.

**Comment by joehandzik** [comment]
Some issues:

1. I couldn't find a packaged deb for the python3 version of pywbem, so I had to use pip3 to pull it in. I used pip3 for pyudev too, but never checked for that package.
2. 'make' and 'make check' work, but when I install, lsmcli doesn't work. Neither does plugin_test.py...they both error out, saying that they cannot find the 'lsm' module. 

I should have another chance to try tomorrow. What should I look for? This is Debian Jessie, fwiw.

**Comment by cathay4t** [comment]
@tasleson The 0.9.0 pywbem rpm SPEC for Fedora could found in https://bugzilla.redhat.com/show_bug.cgi?id=1375892

**Comment by cathay4t** [comment]
@tasleson The pywbem upstream moved 'Error' class from `cim_http.py` to `exceptions.py` file, which cause the issue you mention for `pywbem.cim_http.HTTPError` vs `pywbem.cim_http.Error`.

But actually, we should use `pywbem.Error` in the first place.

**Comment by cathay4t** [comment]
And I added three patches in https://github.com/cathay4t/libstoragemgmt/commits/tony_fix for trivial issues I found.

**Comment by cathay4t** [comment]
I am still waiting a MegaRAID server with Fedora 24 to test python3 MegaRAID plugin, and about to take [PTO][1] from tomorrow(Sep 15 2016) to weekend.
Will provide the test results on MegaRAID python3 plugin next week.

[1]: https://en.wikipedia.org/wiki/Mid-Autumn_Festival

**Comment by joehandzik** [comment]
So on both a 'lsmcli ls' and 'lsmcli ld' with python3, I get back:

PLUGIN_BUG(2): Got unexpected error sequence item 0: expected str instance, bytes found

I'll investigate, but if someone knows off the top of their head...

**Comment by tasleson** [comment]
@joehandzik Guess I haven't seen that issue, but I typically don't use alias's

**Comment by joehandzik** [comment]
@tasleson The alias is irrelevant, problem still exists when using 'lsmcli list --type disks'. I'll have time to look into it today.

**Comment by tasleson** [comment]
@joehandzik  Which plugin were you using, simulator or other?

**Comment by joehandzik** [comment]
The hpsa plugin. I suspect somthing in utils.py.

**Comment by joehandzik** [comment]
@tasleson Confirmed that utils.py is one way to fix this...adding the universal_newlines=True option to the Popen call makes python3 work so far. Should be backwards compatible with 2.x.

For reference: 
Documentation on the argument for python 3 - https://docs.python.org/3/library/subprocess.html#frequently-used-arguments
Proof that the option exists for Popen in python 2 - https://docs.python.org/2/library/subprocess.html#popen-constructor

**Comment by joehandzik** [comment]
@tasleson See PR #209. If it works for you, you can pull in as a separate PR or cherry-pick the patch, whatever works best for you.

**Comment by tasleson** [comment]
@joehandzik  Would it be possible to get the unit test node up and running at your location to automate the hpsa plugin?

**Comment by joehandzik** [comment]
@tasleson It's up! Not 100% sure if I've chewed through all the false positive failures though...that was what #205 was intended to fix.

**Comment by tasleson** [comment]
@joehandzik I will rebase this branch hopefully yet today and we will see if we make it through the unit test with your node.

**Comment by joehandzik** [comment]
@tasleson Ok, sounds good. Lemme bring the node down again for about half an hour, I realized that there's a firmware update that I want to apply to it...I'll post back here when it's back up again.

**Comment by tasleson** [comment]
@joehandzik Take your time, the unit test queue is @ 4 right now, so we have some time.  I'm looking at adding the ability to display what's in the queue, because I'm at a loss to what's backed up.

**Comment by joehandzik** [comment]
@tasleson It's back up. I'm not sure what's backed up either...

**Comment by tasleson** [comment]
@joehandzik I just checked the server logs.  It appears we got events from git hub for your `wip-lsm-popen` branch at the following times (eastern):

``` text
2016-09-19 15:47:03.209326: pid: 11337: Queuing unit tests for https://github.com/joehandzik/libstoragemgmt.git wip-lsm-popen
2016-09-19 15:47:43.049112: pid: 11337: Queuing unit tests for https://github.com/joehandzik/libstoragemgmt.git wip-lsm-popen
2016-09-19 15:50:19.154558: pid: 11337: Queuing unit tests for https://github.com/joehandzik/libstoragemgmt.git wip-lsm-popen
2016-09-19 16:26:28.079767: pid: 11337: Queuing unit tests for https://github.com/joehandzik/libstoragemgmt.git wip-lsm-popen
```

We should only be getting these events on a pull request, but apparently it looks like I may be doing something incorrectly which is causing quite a bit of unit test churn.  I will dump more debug data to the log to see what is going on.

**Comment by joehandzik** [comment]
@tasleson I think two events would make sense. I actually did open a PR against your fork on accident, but then I closed it and opened a PR against the master. No idea where the extra two came from, though I think you may have initiated one of them?

**Comment by joehandzik** [comment]
@tasleson Took the HPE node down intentionally. Didn't have a python2 dependency, apparently...

**Comment by joehandzik** [comment]
@tasleson Reconnected now.

**Comment by tasleson** [comment]
@joehandzik The dependency was for the python-requests library which I forgot to remove as it's no longer needed.  My local branch has this removed.

**Comment by tasleson** [comment]
@joehandzik I've pushed updates for this branch.  It's in the queue for unit test which will complete sometime after other stuff that is queued up, including Blaine's latest PR :-)

**Comment by cathay4t** [comment]
@tasleson Could you includes these patches in this PR:
* Use pywbem instead of lmiwbem for python3 in rpm SPEC  https://github.com/cathay4t/libstoragemgmt/commit/b0942e266e1b86ef5e50746f7f14f210f0f14dce
* lsmcli: Fix failure when exec lsmcli without any argument. https://github.com/cathay4t/libstoragemgmt/commit/4034d1d6688003d1e1929aa7ae8a287b40578c53
* RPM SPEC: Fix compile on RHEL 7 https://github.com/cathay4t/libstoragemgmt/commit/0fa097367204213dbad2f6ef9e528b6494b62c44
* MegaRAID plugin: Fix disk listing on some dell PERC cards. https://github.com/cathay4t/libstoragemgmt/commit/286afa2089557323ce7de44ed6804c200602e6d8
* MegaRAID plugin: Fix disk VPD83 on Dell PERC.  https://github.com/cathay4t/libstoragemgmt/commit/650da8e5dc8608176d5d06477ca82a7388266926

Tested python3 MegaRAID on Fedora 24. It works well with patches above.

**Comment by joehandzik** [comment]
@tasleson Ok, looks like the failures for me are all rooted at capabilities not being checked.

Failures due to not support on my H240 HBA:

- pools list
- volumes list
- volumes vpd83

Failures due to correctly identified no support (honestly not sure why this one was a failure at first glance):

- test_timeout

I'll see if I can figure out a way around this. Otherwise though, it sure would appear that we're close to having Smart Array failures be legitimate...

**Comment by tasleson** [comment]
@joehandzik @cathay4t The ability to get/set a timeout and list pools are not covered by capability as they are expected to always be present (see class IPlugin) .  At the moment I don't remember how were were working around this with local HBA raid cards.

**Comment by tasleson** [comment]
@cathay4t @joehandzik FYI: The automated test runs py2 then py3.  However, py3 only gets run if py2 runs clean.

**Comment by joehandzik** [comment]
@tasleson I guess in the short-term I can look for a way that would work for Smart Array, and then we can test on a Broadcom HBA to see if it works there too. I think it would probably be *most correct* to add capabilities for everything. What do you think?

**Comment by tasleson** [comment]
@joehandzik If we want to maintain the API compatibility we cannot change everything to a capability.  The stuff that isn't covered by capability that is client facing is:

* time_out_set/time_out_get
* job_status/job_free
* capabilities
* plugin_info
* pools
* systems

As timeout don't make much sense for local HBA (although we could timeout a command) those could be stubbed out to simple dummy store and retrieve a value, yeah I know kind of bogus.

`pools` and `systems` could simply return an empty result if nothing is available or supported for the current configuration.  The unit test could be changed to not fail if we don't get anything returned for `pools` and `systems`.  We shouldn't be testing `volumes` if not supported as that is a capability.

How do these proposed changes sound?

**Comment by joehandzik** [comment]
@tasleson We could stub out time_out for sure. Probably need to do as you suggested for pools and return empty and/or rewrite the unit tests to allow for the possibility that volumes or pools are not present. Similar for volume vpd83. That'll cover it for me! I'll aim to have something today.

**Comment by tasleson** [comment]
@joehandzik I just modified the plugin test and added fake timeout support to hpsa, lets see how this turns out.  Unit test running.

**Comment by joehandzik** [comment]
@tasleson Looks like a success! :)

**Comment by tasleson** [comment]
@joehandzik Yep.

I still need to fix some differences up for pywbem and integrate changes from Gris for megaraid, but we are getting close :-)

File: plugin/smispy/WBEM.py [hunk 1]
```diff
@@ -15,7 +15,38 @@
...
+    except ImportError:
+        from pywbem.cim_http import Error
+    try:
+        from pywbem import AuthError
+    except ImportError:
+        from pywbem.cim_http import AuthError
+else:
+    from lsm.plugin.smispy.lmiwbem_wrap import Error
+    from lsm.plugin.smispy.lmiwbem_wrap import AuthError
+
```
Comment (cathay4t) [inline plugin/smispy/WBEM.py]: Please remove the new blank line at EOF.

Comment (tasleson) [inline plugin/smispy/WBEM.py]: Done

File: packaging/libstoragemgmt.spec.in [hunk 2]
```diff
@@ -200,8 +200,8 @@ libraries.
 Summary:        Files for SMI-S generic array support for %{libstoragemgmt}
 Group:          System Environment/Libraries
 %if 0%{?with_python3}
-BuildRequires:  python3-lmiwbem
-Requires:       python3-lmiwbem
+BuildRequires:  python3-pywbem
+Requires:       python3-pywbem
```
Comment (tasleson) [inline packaging/libstoragemgmt.spec.in]: Long term we will be moving to pywbem only.  Short term we will be leveraging lmiwbem until a py3 supported version of pywbem is available in fedora.

**Comment by cathay4t** [comment]
@tasleson Please include patch https://github.com/cathay4t/libstoragemgmt/commit/ccb382f66b0ce3c72cdac86829b8b566aa04b0be into this PR. Tested by me and Joe.

**Comment by tasleson** [comment]
@cathay4t Do we just want to commit this change to master and have me re-base this branch instead?  Not sure how long this branch will be outstanding as I'm still waiting to hear back on test results.

**Comment by cathay4t** [comment]
@tasleson Python2 in debian does not have any issue. Patch cathay4t@ccb382f is only for python3 in debian, so I am afraid it should be included in this PR.

**Comment by tasleson** [comment]
@cathay4t Patch cathay4t@ccb382f  added to PR

**Comment by tasleson** [comment]
@cathay4t @joehandzik  I would like to wrap this up and get this merged, preferably by the end of the week.  I believe hpsa has been tested?  What about megaraid?

**Comment by joehandzik** [comment]
hpsa has been tested and looks good based on my ad-hoc testing and the CI tests that ran against my DL380. I'm not sure how much runtime megaraid has.

**Comment by tasleson** [comment]
@joehandzik looks like your test node went off-line on: 2016-10-14 00:25:39.892955 from the server perspective, any idea why missing?

**Comment by joehandzik** [comment]
@tasleson I had noticed that it went offline, but had not gotten a chance to debug that yet. I think this has happened once before, the server is actually locking up in some fashion (no idea why, this isn't happening on my box that has Centos installed). I'll go restart now, and work on getting the BIOS, NICs, etc updated all the way to remove those as possible issues. From the /var/log/messages logs that I had looked at the last time this happened, it wasn't a storage issue.

**Comment by tasleson** [comment]
@joehandzik Looks like the solution is to install CentOS :bowtie:

**Comment by tasleson** [comment]
I just added another test node for ontap testing (need to ensure it comes up ready when rebooted).  Once we get megaraid up on a test node we will have automated coverage for all :+1:

**Comment by cathay4t** [comment]
@tasleson Could you include patch https://github.com/cathay4t/libstoragemgmt/commit/d560c0d68f31bfda34408c68e36bdb435687dc7e in this PR?
For the MegaRAID part, I am still waiting beaker/lab to install an Fedora megaraid server for me.
If that passes, are we ready for merge?

**Comment by tasleson** [comment]
@cathay4t I rebased this branch and added in 16a830501862de610c1e28cce8543d7e51d8515a as requested.  Once we have megaraid tested we are ready to merge this!

**Comment by cathay4t** [comment]
MegaRAID plugin in python3 and python2 mode are both OK. Merged

**Pull request merged**